services/upgrader/upgrader_linux.go

```diff
@@ -0,0 +1,332 @@
+		return err
+	}
+
+	filename, err := download(getCurrentVersionTgzUrl(), 1024*1024*300)
+	if err != nil {
+		if filename != "" {
+			os.Remove(filename)
+		}
+		upgradeError = fmt.Errorf("error downloading the new version (percentage: %d)", upgradePercentage)
+		mlog.Error("Unable to download the mattermost server version", mlog.Int64("percentage", upgradePercentage), mlog.String("url", getCurrentVersionTgzUrl()), mlog.Err(err))
+		upgradePercentage = 0
+		return err
+	}
+	defer os.Remove(filename)
+	sigfilename, err := download(getCurrentVersionTgzUrl()+".sig", 1024)
+	if err != nil {
+		if sigfilename != "" {
+			os.Remove(sigfilename)
+		}
+		upgradeError = errors.New("error downloading the new version signature file")
+		mlog.Error("Unable to download the mattermost server version signature file", mlog.String("url", getCurrentVersionTgzUrl()+".sig"), mlog.Err(err))
+		upgradePercentage = 0
+		return err
+	}
+	defer os.Remove(sigfilename)
+
+	err = verifySignature(filename, sigfilename, mattermostBuildPublicKey)
+	if err != nil {
+		upgradePercentage = 0
+		upgradeError = errors.New("Unable to verify downloaded file signature")
+		mlog.Error("Unable to verify downloaded file signature", mlog.Err(err))
+		return err
+	}
+
+	err = extractBinary(executablePath, filename)
+	if err != nil {
+		upgradePercentage = 0
+		upgradeError = err
+		mlog.Error("Unable to extract the binary from the downloaded file", mlog.Err(err))
+		return err
+	}
+	upgradePercentage = 100
+	return nil
+}
+
+func UpgradeToE0Status() (int64, error) {
+	return upgradePercentage, upgradeError
+}
+
+func download(url string, limit int64) (string, error) {
+	resp, err := http.Get(url)
+	if err != nil {
+		return "", err
+	}
+	defer resp.Body.Close()
+
+	out, err := ioutil.TempFile("", "*_mattermost.tar.gz")
+	if err != nil {
+		return "", err
+	}
+	defer out.Close()
+
+	counter := &writeCounter{total: resp.ContentLength}
+	_, err = io.Copy(out, io.TeeReader(&io.LimitedReader{R: resp.Body, N: limit}, counter))
+	return out.Name(), err
+}
+
+func getFilePermissionsOrDefault(filename string, def os.FileMode) os.FileMode {
+	file, err := os.Open(filename)
+	if err != nil {
+		mlog.Error("Unable to get the file permissions", mlog.String("filename", filename), mlog.Err(err))
+		return def
+	}
+	fileStats, err := file.Stat()
+	if err != nil {
+		mlog.Error("Unable to get the file permissions", mlog.String("filename", filename), mlog.Err(err))
+		return def
+	}
+	return fileStats.Mode()
+}
+
+func extractBinary(executablePath string, filename string) error {
+	gzipStream, err := os.Open(filename)
+	if err != nil {
+		return err
+	}
+
+	uncompressedStream, err := gzip.NewReader(gzipStream)
+	if err != nil {
+		return err
+	}
+
+	tarReader := tar.NewReader(uncompressedStream)
+
+	for {
+		header, err := tarReader.Next()
+
+		if err == io.EOF {
+			return errors.New("Unable to find mattermost binary in the downloaded version")
```

Comment:
```suggestion
			return errors.New("Unable to find the Mattermost binary in the downloaded version")
```

What does "version" mean here? Would "file" be better? "Resource"? ".tar file"?

Reply:
It is talking about the Mattermost executable file (the binary) in the downloaded .tar.gz file.